File: app/src/main/java/com/github/anrimian/musicplayer/di/app/library/genres/GenresModule.java
package com.github.anrimian.musicplayer.di.app.library.genres;

import com.github.anrimian.musicplayer.domain.interactors.library.LibraryGenresInteractor;
import com.github.anrimian.musicplayer.domain.repositories.EditorRepository;
import com.github.anrimian.musicplayer.domain.repositories.LibraryRepository;
import com.github.anrimian.musicplayer.domain.repositories.SettingsRepository;
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser;
import com.github.anrimian.musicplayer.ui.library.genres.list.GenresListPresenter;

import javax.annotation.Nonnull;
import javax.inject.Named;

import dagger.Module;
import dagger.Provides;
import io.reactivex.rxjava3.core.Scheduler;

import static com.github.anrimian.musicplayer.di.app.SchedulerModule.UI_SCHEDULER;

@Module
public class GenresModule {

    @Provides
    @Nonnull
    LibraryGenresInteractor libraryGenresInteractor(EditorRepository editorRepository,
                                                    LibraryRepository repository,
                                                    SettingsRepository settingsRepository) {
        return new LibraryGenresInteractor(editorRepository, repository, settingsRepository);
    }

    @Provides
    @Nonnull
    GenresListPresenter genreListPresenter(LibraryGenresInteractor interactor,
                                           ErrorParser errorParser,
                                           @Named(UI_SCHEDULER) Scheduler uiScheduler) {
        return new GenresListPresenter(interactor, errorParser, uiScheduler);
    }
}

File: app/src/main/java/com/github/anrimian/musicplayer/di/app/library/genres/GenresModule.kt
package com.github.anrimian.musicplayer.di.app.library.genres

import com.github.anrimian.musicplayer.di.app.SchedulerModule
import com.github.anrimian.musicplayer.domain.interactors.library.LibraryGenresInteractor
import com.github.anrimian.musicplayer.domain.interactors.player.LibraryPlayerInteractor
import com.github.anrimian.musicplayer.domain.interactors.playlists.PlayListsInteractor
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser
import com.github.anrimian.musicplayer.ui.library.genres.list.GenresListPresenter
import dagger.Module
import dagger.Provides
import io.reactivex.rxjava3.core.Scheduler
import javax.inject.Named

@Module
class GenresModule {
    
    @Provides
    fun genreListPresenter(
        interactor: LibraryGenresInteractor,
        playerInteractor: LibraryPlayerInteractor,
        playListsInteractor: PlayListsInteractor,
        errorParser: ErrorParser,
        @Named(SchedulerModule.UI_SCHEDULER) uiScheduler: Scheduler
    ) = GenresListPresenter(
        interactor,
        playerInteractor,
        playListsInteractor,
        errorParser,
        uiScheduler
    )
    
}
